ddtrace/contrib/anthropic/patch.py

```diff
@@ -0,0 +1,164 @@
+    return api_key
+
+
+@with_traced_module
+def traced_chat_model_generate(anthropic, pin, func, instance, args, kwargs):
+    chat_messages = get_argument_value(args, kwargs, 0, "messages")
+    integration = anthropic._datadog_integration
+
+    span = integration.trace(
+        pin,
+        "%s.%s" % (instance.__module__, instance.__class__.__name__),
+        submit_to_llmobs=True,
+        interface_type="chat_model",
+        provider="anthropic",
+        model=kwargs.get("model", ""),
+        api_key=_extract_api_key(instance),
+    )
+    chat_completions = None
+    try:
+        for message_set_idx, message_set in enumerate(chat_messages):
+            if isinstance(message_set, dict):
+                if isinstance(message_set.get("content", None), str):
+                    span.set_tag_str(
+                        "anthropic.request.messages.%d.0.content" % (message_set_idx),
+                        integration.trunc(message_set.get("content", "")),
+                    )
+                    span.set_tag_str(
+                        "anthropic.request.messages.%d.0.type" % (message_set_idx),
+                        "text",
+                    )
+                elif isinstance(message_set.get("content", None), list):
+                    for message_idx, message in enumerate(message_set.get("content", [])):
+                        if integration.is_pc_sampled_span(span):
+                            if message.get("type", None) == "text":
+                                span.set_tag_str(
+                                    "anthropic.request.messages.%d.%d.content" % (message_set_idx, message_idx),
+                                    integration.trunc(str(message.get("text", ""))),
+                                )
+                            elif message.get("type", None) == "image":
+                                span.set_tag_str(
+                                    "anthropic.request.messages.%d.%d.content" % (message_set_idx, message_idx),
+                                    "([IMAGE DETECTED])",
+                                )
+
+                        span.set_tag_str(
+                            "anthropic.request.messages.%d.%d.type" % (message_set_idx, message_idx),
+                            message.get("type", "text"),
+                        )
+        params_to_tag = {k: v for k, v in kwargs.items() if k != "messages"}
+        span.set_tag_str("anthropic.request.parameters", json.dumps(params_to_tag))
+
+        chat_completions = func(*args, **kwargs)
+        if isinstance(chat_completions, anthropic.Stream) or isinstance(
+            chat_completions, anthropic.lib.streaming._messages.MessageStreamManager
+        ):
+            return handle_stream_response(integration, chat_completions, args, kwargs, span)
+        else:
+            return handle_non_stream_response(integration, chat_completions, chat_messages, span)
```

Comment:
For #2, I moved the finally block outside the  `traced_chat_model_generate()` func since we probably don't want to be closing the span for a stream (I guess I could put an if check for a stream, and leave it open if it is)

Reply:
1. I'm good with leaving the `span.finish()` in the  stream response handler
2. Sorry about the back and forth, we can rename the handlers back to `handle_{}_...`. I didn't realize we were returning the traced streamed response, it makes more sense to go with what you originally named the methods
3. Can we make `handle_non_streamed_response()` not return anything and just perform the tagging? This way we can keep the `span.finish()` in the original `traced_chat_model_generate()`.